pkg/export/secrets/watch.go

```diff
@@ -0,0 +1,241 @@
+// You may obtain a copy of the License at
+//
+//	https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package secrets
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"sync"
+	"time"
+
+	"github.com/go-kit/log"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/kubernetes"
+)
+
+// WatchSPConfig configures access to the Kubernetes API server.
+// TODO(TheSpiritXIII): https://github.com/GoogleCloudPlatform/prometheus-engine/issues/867
+type WatchSPConfig struct {
+	ClientConfig
+}
+
+func (c *WatchSPConfig) newProvider(ctx context.Context, opts ProviderOptions) (*watchProvider, error) {
+	client, err := c.ClientConfig.client()
+	if err != nil {
+		return nil, err
+	}
+	return newWatchProvider(ctx, opts.Logger, client)
+}
+
+type watcher struct {
+	mu       sync.Mutex
+	w        watch.Interface
+	refCount uint
+	s        *corev1.Secret
+}
+
+func newWatcher(ctx context.Context, logger log.Logger, client kubernetes.Interface, config *KubernetesSecretConfig) (*watcher, error) {
+	val := &watcher{
+		refCount: 1,
+	}
+
+	var err error
+	val.w, err = client.CoreV1().Secrets(config.Namespace).Watch(ctx, metav1.ListOptions{
+		FieldSelector:       fields.OneTermEqualSelector(metav1.ObjectNameField, config.Name).String(),
+		AllowWatchBookmarks: false,
+	})
+	if err != nil {
+		return val, fmt.Errorf("unable to watch secret %s/%s: %w", config.Namespace, config.Name, err)
+	}
+
+	// We could wait for the first watch event, but it doesn't notify us if the resource doesn't exist.
+	val.s, err = client.CoreV1().Secrets(config.Namespace).Get(ctx, config.Name, metav1.GetOptions{})
+	if err != nil && !apierrors.IsNotFound(err) {
+		return val, fmt.Errorf("unable to fetch secret %s/%s: %w", config.Namespace, config.Name, err)
+	}
+	go func() {
+		for {
+		channelLoop:
+			for {
+				select {
+				case e, ok := <-val.w.ResultChan():
+					if !ok {
+						break channelLoop
+					}
+					val.update(logger, e)
+				case <-ctx.Done():
+					// Run within separate function so we can use defer.
+					func() {
+						val.mu.Lock()
+						defer val.mu.Unlock()
+						val.w.Stop()
+					}()
+					return
+				}
+			}
+
+			// Run within separate function so we can use defer.
+			finished := func() bool {
+				// Check in case the channel cancelled intentionally.
+				if val.refCount == 0 {
+					return true
+				}
+
+				// Pseudo-arbitrarily jitter the length of the most common scrape interval.
+				time.Sleep(1*time.Second + time.Duration(rand.Intn(30)))
```

Comment:
question: What's the utility of this sleep? And you're sleeping 1s + [0,30) [nanoseconds](https://pkg.go.dev/time#Duration)?

Reply:
Whoops, this was meant to be [1, 31] seconds. This is some jitter because if the Kubernetes API server goes down, we don't want to send potentially dozens of watches at once across all the pods in the collector DaemonSet.